src/creature.cpp

```diff
@@ -229,15 +229,17 @@ void Creature::onWalk()
 
 void Creature::onWalk(Direction& dir)
 {
-	if (hasCondition(CONDITION_DRUNK)) {
-		uint32_t r = uniform_random(0, 20);
-		if (r <= DIRECTION_DIAGONAL_MASK) {
-			if (r < DIRECTION_DIAGONAL_MASK) {
-				dir = static_cast<Direction>(r);
-			}
-			g_game.internalCreatureSay(this, TALKTYPE_MONSTER_SAY, "Hicks!", false);
-		}
+	if (!hasCondition(CONDITION_DRUNK)) {
+		return;
 	}
+
+	uint8_t rand = uniform_random(0, 99);
+	if (rand > getDrunkenness()) {
+		return;
+	}
+
+	dir = static_cast<Direction>(rand % 4);
```

Comment:
If you put a chance of 1%, the drunk player will always "random" step north, because the value of `rand` here will always be 0. You need to re-roll the die to get the direction. Preferably, use the enum values instead of naked integers.

Reply:
I'd like to hear a second opinion on this one (@nekiro @EPuncker).

First of all, why would you set a 1% chance of drunkenness in the first place? Let alone, any value <5-10%? Moreover,
assuming you have a created a drunk condition with a drunkenness' value of 1% and, assuming again, it lasts for at least 200 SQMs, would you be able to tell wether your 2~4 drunk steps moved you to the north? Is it probable that you move 200 SQMs in a single direction?

Also, "rolling the die" again consumes CPU. Is it worth delaying the response time of the server in order to provide a, so called, "more realistic drunk effect" in such an edge case?

As for using enum values, do you think `rand % DIRECTION_DIAGONAL_MASK` would make the method more readable? In my opinion it adds a, rather useless, level of indirection. Are those values going to change in the foreseable future?

Reply:
> First of all, why would you set a 1% chance of drunkenness in the first place?

Well, because it is allowed. I don't know what are reasonable values for this, to be fair.

> Also, "rolling the die" again consumes CPU

I wouldn't even worry about it, as the [Mersenne Twister, available as `std::mt19937`, can generate a random number in 6 cycles](https://github.com/s9w/articles/blob/master/perf%20cpp%20random.md). You could generate a million random numbers and the response time will barely be affected, even using the slower `uniform_int_distribution`.

You can, however, generate a number between 0 and 400, use the last 2 bits for the direction and the high bits for chance, i.e.

```cpp
	uint16_t rand = uniform_random(0, 399);
	if (rand / 4 > getDrunkenness()) {
		return;
	}

	dir = static_cast<Direction>(rand % 4);
```

This is still guaranteed to be uniformly distributed.

> As for using enum values, do you think `rand % DIRECTION_DIAGONAL_MASK` would make the method more readable?

Not more readable, but more semantic. I meant `uniform_random(DIRECTION_NORTH, DIRECTION_WEST)`, but thinking again it doesn't help that much indeed, so dismiss it.

Reply:
Any reason not to trust `uniform_random`?

Reply:
> You can, however, generate a number between 0 and 400, use the last 2 bits for the direction and the high bits for chance, i.e.

I like this solution very much indeed. Let me update the branch.